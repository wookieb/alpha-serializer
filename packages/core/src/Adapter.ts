export interface Adapter<T = any> {

    readonly excludedNormalizations?: string[];
    serialize(data: any): T;

    deserialize(data: T): any;
}