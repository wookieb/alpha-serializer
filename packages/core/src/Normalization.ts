export class Normalization<T = any> {
    constructor(public readonly name: string,
                public readonly clazz: { new(...args: any[]): any },
                public readonly normalizer: NormalizerFunction<T>,
                public readonly denormalizer: DenormalizerFunction<T>) {
        Object.freeze(this);
    }
}

export type NormalizerFunction<T = any> = (value: T) => any;
export type DenormalizerFunction<T = any> = (value: any) => T;