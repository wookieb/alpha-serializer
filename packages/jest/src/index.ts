import {create} from "./create";
import {normalizer} from "alpha-serializer";

module.exports = Object.assign(create(normalizer), {create});