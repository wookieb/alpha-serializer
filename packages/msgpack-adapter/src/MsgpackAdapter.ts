import {Adapter, normalizations} from 'alpha-serializer';

const msgpack = require('msgpack5');

export interface CustomType<T> {
    type: number;
    normalizationName?: string;
    class: { new(...args: any[]): any };
    encode: (value: T) => Buffer;
    decode: (value: Buffer) => T;
}

export interface AdapterOptions {
    forceFloat64?: boolean;
    compatibilityMode?: boolean;
    disableTimestampEncoding?: boolean;
}

export class MsgpackAdapter implements Adapter<Buffer> {
    private msgpack: any;
    private customTypesNormalizations: string[] = [];
    private readonly isDateSerializationEnabled: boolean;

    constructor(options?: AdapterOptions) {
        this.msgpack = msgpack(options);

        this.isDateSerializationEnabled = options ? !options.disableTimestampEncoding : true;
    }

    public serialize(data: any): Buffer {
        return this.msgpack.encode(data).slice();
    }

    public deserialize(input: Buffer): any {
        return this.msgpack.decode(input);
    }

    /**
     * Registers custom encoding and decoding for objects of given class
     *
     * If "normalizationName" gets provided then such normalizations gets disabled
     */
    public registerCustomType(customType: CustomType<any>): this {
        if (customType.normalizationName) {
            this.customTypesNormalizations.push(customType.normalizationName);
        }
        this.msgpack.register(customType.type, customType.class, customType.encode, customType.decode);
        return this;
    }

    get excludedNormalizations() {
        if (this.isDateSerializationEnabled) {
            return this.customTypesNormalizations.concat([normalizations.DATE.name]);
        }
        return this.customTypesNormalizations;
    }
}
