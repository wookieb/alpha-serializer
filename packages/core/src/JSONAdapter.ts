import {Adapter} from "./Adapter";

export class JSONAdapter implements Adapter<string> {
    public serialize(data: any) {
        return JSON.stringify(data);
    }

    public deserialize(data: string): any {
        return JSON.parse(data);
    }
}