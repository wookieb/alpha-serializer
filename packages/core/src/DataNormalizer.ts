import {Normalization, NormalizationInput} from "./Normalization";
import * as is from 'predicates';
import {simpleDenormalizer, simpleNormalizer} from "./normalizerFactory";

export const NORMALIZED_TYPE_KEY = '@type';

export class DataNormalizer {
    private nameToNormalization: Map<string, Normalization> = new Map();
    private clazzToNormalization: Map<new(...args: any[]) => any, Normalization> = new Map();

    constructor(public useProxy: boolean = true) {
    }

    public registerNormalization(normalizationInput: NormalizationInput) {
        const normalization = this.normalizationInputToNormalization(normalizationInput);
        this.nameToNormalization.set(normalization.name, normalization);
        this.clazzToNormalization.set(normalization.clazz, normalization);
    }

    private normalizationInputToNormalization(input: NormalizationInput) {
        is.assert(is.property('clazz', is.func), '"clazz" property is required and has to be a function');
        return new Normalization(
            input.name || input.clazz.name,
            input.clazz,
            input.normalizer ? input.normalizer : simpleNormalizer,
            input.denormalizer ? input.denormalizer : simpleDenormalizer.bind(null, input.clazz)
        );
    }

    public getNormalization(name: string) {
        return this.nameToNormalization.get(name);
    }

    /**
     * Normalizes data using registered normalizers.
     * Some normalizations might be excluded (by adapters for example)
     */
    public normalize(data: any, excludedNormalizations?: string[]): any {
        if (is.primitive(data)) {
            return data;
        }

        if (is.function(data)) {
            return;
        }

        if (is.array(data)) {
            return Array.prototype.map.call(data, (value: any) => this.normalize(value), this);
        }

        const normalization = this.getNormalizationForObject(data);

        if (normalization) {
            if (excludedNormalizations && excludedNormalizations.indexOf(normalization.name) !== -1) {
                return data;
            }
            const normalized = normalization.normalizer(data);
            return {
                [NORMALIZED_TYPE_KEY]: normalization.name,
                value: is.primitive(normalized) ? normalized : this.internalNormalize(normalized)
            };
        }

        return this.internalNormalize(data);
    }

    private internalNormalize(data: any) {
        return this.useProxy ? this.normalizeWithProxy(data) : this.normalizeWithoutProxy(data);
    }

    private normalizeWithProxy(data: any) {
        return new Proxy(
            this.getProxyTargetForNormalizedValue(data),
            this.createProxyHandlerForValue(data),
        );
    }

    private normalizeWithoutProxy(data: any) {
        if (Array.isArray(data)) {
            return data.map((x: any) => this.normalize(x));
        }
        const result: any = {};
        for (const key of Object.keys(data)) {
            result[key] = this.normalize(data[key]);
        }
        return result;
    }

    /**
     * Checks whether given value has any defined normalizations
     * Primitives, functions and arrays are ignored
     *
     * @param value
     * @returns {boolean}
     */
    public hasNormalization(value: any) {
        if (is.primitive(value) || is.function(value) || is.array(value)) {
            return false;
        }

        return is.object(value) && !!this.getNormalizationForObject(value);
    }

    public denormalize(data: any): any {
        if (is.primitive(data)) {
            return data;
        }

        if (is.array(data)) {
            return data.map(this.denormalize, this);
        }

        if (NORMALIZED_TYPE_KEY in data) {
            const normalization = this.getNormalizationForType(data[NORMALIZED_TYPE_KEY]);
            if (normalization) {
                return normalization.denormalizer(this.denormalize(data.value));
            } else {
                throw new Error(`Missing normalization for type ${data[NORMALIZED_TYPE_KEY]}`);
            }
        }

        const newValue: any = {};
        // tslint:disable-next-line: forin
        for (const key in data) {
            newValue[key] = this.denormalize(data[key]);
        }
        return newValue;
    }

    private getProxyTargetForNormalizedValue(value: any) {
        if (Array.isArray(value)) {
            return [];
        }
        return {};
    }

    private createProxyHandlerForValue(value: any) {
        return {
            getPrototypeOf() {
                return Object.getPrototypeOf(value);
            },
            has(target: any, property: string | number) {
                return property in value;
            },
            ownKeys() {
                return Object.getOwnPropertyNames(value);
            },
            getOwnPropertyDescriptor(target: any, property: string | number) {
                const descriptor = Object.getOwnPropertyDescriptor(value, property);
                // make a property configurable
                if (descriptor && !descriptor.configurable) {
                    return {...descriptor, configurable: true};
                }
                return descriptor;
            },
            get: (target: any, property: string | number) => {
                const val = value[property];
                // usually functions are not serialized either so it's safe to return them
                if (is.func(val)) {
                    return val;
                }
                return this.normalize(val);
            },
        };
    }

    private getNormalizationForObject(object: any) {
        const prototype = Object.getPrototypeOf(object);

        if (prototype) {
            return this.clazzToNormalization.get(prototype.constructor);
        }
    }

    private getNormalizationForType(type: string) {
        return this.nameToNormalization.get(type);
    }
}