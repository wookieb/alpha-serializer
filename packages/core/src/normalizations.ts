import {Normalization} from "./Normalization";

export const MAP = new Normalization<Map<any, any>>(
    'Map',
    Map,
    (data: Map<any, any>) => {
        return Array.from(data.entries());
    },
    (data: any[]) => {
        return new Map(data);
    },
);

export const SET = new Normalization<Set<any>>(
    'Set',
    Set,
    (data: Set<any>) => {
        return Array.from(data);
    },
    (data: any[]) => {
        return new Set(data);
    },
);

export const DATE = new Normalization<Date>(
    'Date',
    Date,
    (data: Date) => {
        return data.toISOString();
    },
    (data: string) => {
        return new Date(data);
    },
);