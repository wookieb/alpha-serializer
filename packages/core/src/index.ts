import {Serializable} from "./decorators/Serializable";
import {StandardNormalizer} from "./StandardNormalizer";
import * as normalizations from './normalizations';
import {Serializer} from "./Serializer";
import {JSONAdapter} from "./JSONAdapter";

export * from './DataNormalizer';
export * from './Normalization';
export * from './Serializer';
export * from './JSONAdapter';
export * from './normalizerFactory';
export * from './Adapter';
export * from './SerializerInterface';

export {Serializable, StandardNormalizer, normalizations};

export const normalizer = new StandardNormalizer();
Serializable.useDataNormalizer(normalizer);

export const serializer = new Serializer(new JSONAdapter(), normalizer);
export const serialize = serializer.serialize.bind(serializer);
export const deserialize = serializer.deserialize.bind(serializer);

export const registerNormalization = normalizer.registerNormalization.bind(normalizer);