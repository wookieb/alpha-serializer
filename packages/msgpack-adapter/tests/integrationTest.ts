import {Serializer, globalNormalizer} from 'alpha-serializer';
import {MsgpackAdapter} from "../src";
import each from 'jest-each';

describe('integration', () => {

    const values = [
        [new Map<string, any>([['key', new Date(1000)], ['key2', 100], ['key3', null]])],
        [new Set<any>([new Date(1001), 'string', null, 1000])],
        [new Date(1010)],
        [{
            foo: 'bar',
            prop: 1,
            date: new Date(1002),
        }],
        [[new Date(1003), 'foo', 'bar', null, 100]],
    ];

    function createTest(serializer: Serializer) {
        return (value: any) => {
            const serialized = serializer.serialize(value);
            const deserialized = serializer.deserialize(serialized);

            expect(serialized).toMatchSnapshot();
            expect(deserialized).toEqual(value);
        };
    }

    describe('simple adapter', () => {
        const serializer = new Serializer(globalNormalizer, new MsgpackAdapter());
        each(values).test('case %p', createTest(serializer));
    });

    describe('adapter without timestamp encoding', () => {
        const serializer = new Serializer(globalNormalizer, new MsgpackAdapter({
            disableTimestampEncoding: true,
        }));
        each(values).test('case %p', createTest(serializer));
    });
});
