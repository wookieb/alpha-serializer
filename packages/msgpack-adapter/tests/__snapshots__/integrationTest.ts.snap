// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration adapter without timestamp encoding case [1970-01-01T00:00:01.003Z, "foo", "bar", null, 100] 1`] = `
Object {
  "data": Array [
    149,
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    48,
    51,
    90,
    163,
    102,
    111,
    111,
    163,
    98,
    97,
    114,
    192,
    100,
  ],
  "type": "Buffer",
}
`;

exports[`integration adapter without timestamp encoding case 1970-01-01T00:00:01.010Z 1`] = `
Object {
  "data": Array [
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    49,
    48,
    90,
  ],
  "type": "Buffer",
}
`;

exports[`integration simple adapter case [1970-01-01T00:00:01.003Z, "foo", "bar", null, 100] 1`] = `
Object {
  "data": Array [
    149,
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    48,
    51,
    90,
    163,
    102,
    111,
    111,
    163,
    98,
    97,
    114,
    192,
    100,
  ],
  "type": "Buffer",
}
`;

exports[`integration simple adapter case {"date": 1970-01-01T00:00:01.002Z, "foo": "bar", "prop": 1} 1`] = `
Object {
  "data": Array [
    131,
    163,
    102,
    111,
    111,
    163,
    98,
    97,
    114,
    164,
    112,
    114,
    111,
    112,
    1,
    164,
    100,
    97,
    116,
    101,
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    48,
    50,
    90,
  ],
  "type": "Buffer",
}
`;

exports[`integration simple adapter case 1970-01-01T00:00:01.010Z 1`] = `
Object {
  "data": Array [
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    49,
    48,
    90,
  ],
  "type": "Buffer",
}
`;

exports[`integration simple adapter case Map {"key" => 1970-01-01T00:00:01.000Z, "key2" => 100, "key3" => null} 1`] = `
Object {
  "data": Array [
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    163,
    77,
    97,
    112,
    165,
    118,
    97,
    108,
    117,
    101,
    147,
    146,
    163,
    107,
    101,
    121,
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    48,
    48,
    90,
    146,
    164,
    107,
    101,
    121,
    50,
    100,
    146,
    164,
    107,
    101,
    121,
    51,
    192,
  ],
  "type": "Buffer",
}
`;

exports[`integration simple adapter case Set {1970-01-01T00:00:01.001Z, "string", null, 1000} 1`] = `
Object {
  "data": Array [
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    163,
    83,
    101,
    116,
    165,
    118,
    97,
    108,
    117,
    101,
    148,
    130,
    165,
    64,
    116,
    121,
    112,
    101,
    164,
    68,
    97,
    116,
    101,
    165,
    118,
    97,
    108,
    117,
    101,
    184,
    49,
    57,
    55,
    48,
    45,
    48,
    49,
    45,
    48,
    49,
    84,
    48,
    48,
    58,
    48,
    48,
    58,
    48,
    49,
    46,
    48,
    48,
    49,
    90,
    166,
    115,
    116,
    114,
    105,
    110,
    103,
    192,
    205,
    3,
    232,
  ],
  "type": "Buffer",
}
`;
